Distance-2 Graph Coloring

william76 opened the pull request:
Put in a lot of cleanup on the D2 coloring.
- I have parked the team-policy variations of this on a branch on my fork to look at later if we want to revive that and dig into it later.
- Distance2 coloring options:
  - `COLORING_D2_SERIAL` - runs the serial version.
  - `COLORING_D2_MATRIX_SQUARED` - runs the variant that squares the matrix and then computes Distance-1 coloring on the result.
  - `COLORNG_D2_VB` - vertex based D2 coloring using greedy approach.
  - `COLORING_D2_VB_BIT` - use of a bit-vector in place of the forbidden array to improve performance.
  - `COLORING_D2_VB_BIT_EF` - Adds basic edge-filtering in.

### Things that might affect backwards compatibility:
- The file `src/graph/KokkosGraph_graph_color.hpp`  is now `src/graph/KokkosGraph_GraphColor.hpp` to more closely match the camelcase style that I see on other files such as `KokkosKernels_HashmapAccumulator.hpp` etc.

### TODO:
- [x] Spot check on RHEL
- [x] Spot check on CUDA
- [x] Add in `src/graph/KokkosGraph_graph_color.hpp` file that just includes the new file for backwards compatibility.

william76 commented:
### Spot Check Results

#### For RHEL7
```
$ ./do-spot-check.sh
--------------------------------------------------------------------------------
-
-           K O K K O S   K E R N E L S   S P O T   C H E C K
-
--------------------------------------------------------------------------------
Kokkos
- URL...: git@github.com:william76/kokkos.git
- Branch: develop
- SHA1..: 7af7f80639337985ddc41eb0814f115f25f51d30

KokkosKernels
- URL...: git@github.com:william76/kokkos-kernels.git
- Branch: Distance2GraphColor-WIP-v002
- SHA1..: 396a34f0bdd754be0241c028f93106219c558d67

Spot Check Command
/home/wcmclen/dev/kk-dev/source/KokkosKernels/scripts/test_all_sandia \
    --spot-check \
    --kokkoskernels-path=/home/wcmclen/dev/kk-dev/source/KokkosKernels \
    --kokkos-path=/home/wcmclen/dev/kk-dev/source/Kokkos \
    --arch="" \
    --skip-hwloc \
    gcc  \
    --num=2

ModuleCmd_Load.c(213):ERROR:105: Unable to locate a modulefile for 'sems-git'
Running on machine: sems
Going to test compilers:  gcc/5.3.0 gcc/7.2.0
Testing compiler gcc/5.3.0
Testing compiler gcc/7.2.0
  Starting job gcc-5.3.0-Serial-release
  Starting job gcc-7.2.0-Serial-release
  PASSED gcc-7.2.0-Serial-release
  PASSED gcc-5.3.0-Serial-release
#######################################################
PASSED TESTS
#######################################################
gcc-5.3.0-Serial-release build_time=160 run_time=526
gcc-7.2.0-Serial-release build_time=156 run_time=495
#######################################################
FAILED TESTS
#######################################################
```

#### On White (CUDA)
```
$ ./do-spot-check.sh
--------------------------------------------------------------------------------
-
-           K O K K O S   K E R N E L S   S P O T   C H E C K
-
--------------------------------------------------------------------------------

Kokkos
- URL...: git@github.com:william76/kokkos.git
- Branch: master
- SHA1..: 9614f72c75aa2131d56900511e5eebae54a7bd8b

KokkosKernels
- URL...: git@github.com:william76/kokkos-kernels.git
- Branch: Distance2GraphColor-WIP-v002
- SHA1..: b46623036e64d7de0708e8d15ae294d64e0a93c6

Spot Check Command
/home/wcmclen/dev/kk-dev/source/KokkosKernels/scripts/test_all_sandia \
    --spot-check \
    --kokkoskernels-path=/home/wcmclen/dev/kk-dev/source/KokkosKernels \
    --kokkos-path=/home/wcmclen/dev/kk-dev/source/Kokkos \
    --arch="Power8,Kepler37" \
    --skip-hwloc \
    cuda \
    gcc \
    --num=2

Running on machine: white
Going to test compilers:  cuda/9.2.88 gcc/6.4.0 gcc/7.2.0
Testing compiler cuda/9.2.88
Testing compiler gcc/6.4.0
  Starting job cuda-9.2.88-Cuda_OpenMP-release
  PASSED cuda-9.2.88-Cuda_OpenMP-release
Testing compiler gcc/7.2.0
  Starting job gcc-6.4.0-OpenMP_Serial-release
  PASSED gcc-6.4.0-OpenMP_Serial-release
  Starting job gcc-7.2.0-OpenMP-release
  PASSED gcc-7.2.0-OpenMP-release
  Starting job gcc-7.2.0-Serial-release
  PASSED gcc-7.2.0-Serial-release
  Starting job gcc-7.2.0-OpenMP_Serial-release
  PASSED gcc-7.2.0-OpenMP_Serial-release
#######################################################
PASSED TESTS
#######################################################
cuda-9.2.88-Cuda_OpenMP-release build_time=875 run_time=499
gcc-6.4.0-OpenMP_Serial-release build_time=354 run_time=909
gcc-7.2.0-OpenMP-release build_time=227 run_time=230
gcc-7.2.0-OpenMP_Serial-release build_time=358 run_time=896
gcc-7.2.0-Serial-release build_time=196 run_time=635
#######################################################
FAILED TESTS
#######################################################

--------------------------------------------------------------------------------
- Elapsed time: 5189 seconds.
0m0.000s 0m0.016s
903m58.108s 16m21.893s
--------------------------------------------------------------------------------
```

william76 commented:
@ndellingwood 
Don't merge this just yet... I found a couple things to clean up this morning. I'm running spot checks on it in a different branch and will merge that into this one provided those pass. I'll post the updated spot check results here when it's done.

srajama1 commented:
I would like to understand these changes before they get merged. Give me some time.

william76 commented:
### Spot Checks

#### RHEL7
```
--------------------------------------------------------------------------------
-
-           K O K K O S   K E R N E L S   S P O T   C H E C K
-
--------------------------------------------------------------------------------

Kokkos
- URL...: git@github.com:william76/kokkos.git
- Branch: develop
- SHA1..: 534b7f32597b885097b91873edf74be02079a5dc

KokkosKernels
- URL...: git@github.com:william76/kokkos-kernels.git
- Branch: Distance2GraphColor-WIP-v003
- SHA1..: 0417948cc5c91d4b004b31345eca49d45004dabc

Spot Check Command
/home/wcmclen/dev/kk-dev/source/KokkosKernels/scripts/test_all_sandia \
    --spot-check \
    --kokkoskernels-path=/home/wcmclen/dev/kk-dev/source/KokkosKernels \
    --kokkos-path=/home/wcmclen/dev/kk-dev/source/Kokkos \
    --arch="" \
    --skip-hwloc \
    gcc  \
    --num=3

Running on machine: sems
Going to test compilers:  gcc/5.3.0 gcc/7.2.0
Testing compiler gcc/5.3.0
Testing compiler gcc/7.2.0
  Starting job gcc-5.3.0-Serial-release
  Starting job gcc-7.2.0-Serial-release
  PASSED gcc-7.2.0-Serial-release
  PASSED gcc-5.3.0-Serial-release
#######################################################
PASSED TESTS
#######################################################
gcc-5.3.0-Serial-release build_time=148 run_time=579
gcc-7.2.0-Serial-release build_time=143 run_time=501
#######################################################
FAILED TESTS
#######################################################

--------------------------------------------------------------------------------
- Elapsed time: 728 seconds.
0m0.009s 0m0.007s
116m38.921s 6m44.494s
--------------------------------------------------------------------------------
```

#### White (7G)
```
--------------------------------------------------------------------------------
-
-           K O K K O S   K E R N E L S   S P O T   C H E C K
-
--------------------------------------------------------------------------------

Kokkos
- URL...: git@github.com:william76/kokkos.git
- Branch: develop
- SHA1..: 7af7f80639337985ddc41eb0814f115f25f51d30

KokkosKernels
- URL...: git@github.com:william76/kokkos-kernels.git
- Branch: Distance2GraphColor-WIP-v003
- SHA1..: 0417948cc5c91d4b004b31345eca49d45004dabc

Spot Check Command
/home/wcmclen/dev/kk-dev/source/KokkosKernels/scripts/test_all_sandia \
    --spot-check \
    --kokkoskernels-path=/home/wcmclen/dev/kk-dev/source/KokkosKernels \
    --kokkos-path=/home/wcmclen/dev/kk-dev/source/Kokkos \
    --arch="Power8,Pascal60" \
    --skip-hwloc \
    cuda \
    gcc \
    --num=2

Running on machine: white
Going to test compilers:  cuda/9.2.88 gcc/6.4.0 gcc/7.2.0
Testing compiler cuda/9.2.88
Testing compiler gcc/6.4.0
  Starting job cuda-9.2.88-Cuda_OpenMP-release
  PASSED cuda-9.2.88-Cuda_OpenMP-release
Testing compiler gcc/7.2.0
  Starting job gcc-6.4.0-OpenMP_Serial-release
  PASSED gcc-6.4.0-OpenMP_Serial-release
  Starting job gcc-7.2.0-OpenMP-release
  PASSED gcc-7.2.0-OpenMP-release
  Starting job gcc-7.2.0-Serial-release
  PASSED gcc-7.2.0-Serial-release
  Starting job gcc-7.2.0-OpenMP_Serial-release
  PASSED gcc-7.2.0-OpenMP_Serial-release
#######################################################
PASSED TESTS
#######################################################
cuda-9.2.88-Cuda_OpenMP-release build_time=909 run_time=373
gcc-6.4.0-OpenMP_Serial-release build_time=367 run_time=904
gcc-7.2.0-OpenMP-release build_time=224 run_time=228
gcc-7.2.0-OpenMP_Serial-release build_time=324 run_time=901
gcc-7.2.0-Serial-release build_time=210 run_time=636
#######################################################
FAILED TESTS
#######################################################

--------------------------------------------------------------------------------
- Elapsed time: 5083 seconds.
0m0.016s 0m0.000s
916m50.238s 16m3.683s
--------------------------------------------------------------------------------
```

ndellingwood commented:
@srajama1 may I add you as a reviewer for this PR?

srajama1 commented:
I will review this.

srajama1 commented on src/graph/KokkosGraph_GraphColor.hpp:
```diff
@@ -0,0 +1,209 @@
+}
+
+
+// Distance 2 graph coloring -- serial only -- should be replaced by the Distance2GraphColoring
```
Can we remove d2 from here ?

srajama1 commented on src/graph/KokkosGraph_GraphColor.hpp:
```diff
@@ -0,0 +1,209 @@
+#include "KokkosGraph_GraphColorHandle.hpp"
+#include "KokkosKernels_Utils.hpp"
+
+namespace KokkosGraph{
```
Let us call this distance-1 color to be consistent with the distance-2 naming.

srajama1 commented on src/graph/KokkosGraph_GraphColorHandle.hpp:
```diff
@@ -60,8 +63,6 @@ enum ColoringAlgorithm { COLORING_DEFAULT,
                          COLORING_EB,                         // Edge Based Coloring
                          COLORING_SERIAL2,
                          COLORING_SPGEMM,
```
This is distance2 right ? Can we move this to ditance2. What is SERIAL2, Can we have more descriptive for name ?

william76 replied:
The `COLORING_SPGEMM` should be the Distance-2 Coloring using SPGEMM + D1 Coloring. I don't think that's in the D1 coloring handle anymore so I'll remove this.

The `COLORING_SERIAL2` is still in the Distance1 coloring handle as an artifact from a while back. I think it's used in the SPGEMM code in a couple places so it's not quite so trivial to cut that out.  I'd like to move this over fully to the Distance2 handle. Maybe it might be good to file an issue and do that part as a separate PR so we can get the Distance2 coloring in since there are some bigger changes with interfaces coming in with this?

srajama1 commented on src/graph/KokkosGraph_GraphColorHandle.hpp:
```diff
@@ -483,11 +498,12 @@ class GraphColoringHandle
       typedef Kokkos::RangePolicy<HandleExecSpace> my_exec_space;
 
-      if (
+      if ( 0
```
What is this if (0) ?

srajama1 replied:
Is this a compiler trick ?

william76 replied:
I think we talked about this one last week, I think what this:

```C++
        if ( 0
#if defined( KOKKOS_ENABLE_CUDA )
        || Kokkos::Impl::is_same<Kokkos::Cuda, ExecutionSpace >::value
#endif
         )
```

is saying is that the stuff in the block should only get executed if `KOKKOS_ENABLE_CUDA` is true -and- if the ExecutionSpace is `Kokkos::Cuda`, otherwise it should not get executed. I didn't write this thing so I'm assuming it's correct.

srajama1 commented on src/graph/impl/KokkosGraph_Distance2Color_MatrixSquared_impl.hpp:
```diff
@@ -0,0 +1,238 @@
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Siva Rajamanickam (srajama@sandia.gov)
```
Feel free to change your name ...  I don't need to be the contact for everything

srajama1 commented on src/graph/impl/KokkosGraph_Distance2Color_impl.hpp:
```diff
@@ -40,899 +40,1615 @@
-                          chunkSize_
-                          );
-
-    Kokkos::parallel_for("KokkosGraph::Distance2Color::ColorGraphGreedy", my_exec_space(0, current_vertexListLength_ / chunkSize_ + 1), gc);
```
Please make sure this is not accidental

william76 replied:
This is fine. It was changed to functorGreedyColorVB to be more consistent with other coloring naming schemes like functorGreedyColorVB_BIT, etc.

srajama1 commented on src/graph/impl/KokkosGraph_Distance2Color_impl.hpp:
```diff
@@ -40,899 +40,1615 @@
+                             const_lno_nnz_view_t     adj_,
+                             const_clno_row_view_t    t_xadj_,
+                             const_clno_nnz_view_t    t_adj_,
+                             color_view
```
Remove this ?

srajama1 commented on src/graph/impl/KokkosGraph_Distance2Color_impl.hpp:
```diff
@@ -40,899 +40,1615 @@
+                             const_lno_nnz_view_t     adj_,
+                             const_clno_row_view_t    t_xadj_,
+                             const_clno_nnz_view_t    t_adj_,
+                             color_view
```
Lot of experimental stuff here ... Do we want to push ?

william76 replied:
@srajama1 This is part of the tight-bound-on-distance2-degree stuff. I've moved the `graph_compute_distance2_degree` function out of `KokkosGraph_Distance2Color.hpp` and into `KokkosGraph_Distance2Color_impl.hpp` so this thing only can get called if you're using `KokkosGraph::Impl` namespace is that ok or would it be better to try and move it into something in an `Experimental` namespace?

srajama1 replied:
If it is not exposed to user and we plan to clean it up soon, I am fine with it. Please file an issue so we don't forget.

william76 commented on perf_test/graph/KokkosGraph_color_d2.cpp:
```diff
@@ -35,556 +35,828 @@
-        }
-      }
 
+template<typename size_type, typename lno_t, typename exec_space, typename hbm_mem_space, typename sbm_mem_space>
```
Correct.  I've cleaned up the file on my branch to remove the multimem stuff.

william76 commented on src/graph/KokkosGraph_Distance2Color.hpp:
```diff
@@ -0,0 +1,309 @@
+    return output;
+}
+
+
```
I've moved these 'helper' functions over into the `KokkosGraph::Impl` namespace and put them in `KokkosGraph_Distance2Color_impl.hpp` on a separate branch.

william76 commented:
I'm doing my updates on a separate branch of my fork to minimize the email spam as I push in commits. Once I get that done I'll merge that branch into this one to update the PR.

william76 commented on src/common/KokkosKernels_Handle.hpp:
```diff
@@ -372,25 +386,35 @@ class KokkosKernelsHandle{
+    // (wcmclen): Should there be a check here to make sure we've created a GC handle before
+    //            handing the pointer out to something? This is disabled for now because it 
+    //            gets thrown in tests run by spot check. Moving forward, we should consider
+    //            whether a "get the handle ptr, then allocate" vs. "only give out the handle ptr
```
I commented that check out on the distance-2 getter as well and referenced this comment.

william76 commented on src/graph/KokkosGraph_GraphColor.hpp:
```diff
@@ -0,0 +1,209 @@
+*/
+#ifndef _KOKKOS_GRAPH_COLOR_HPP
+#define _KOKKOS_GRAPH_COLOR_HPP
+
```
Added in a deprecation warning to the header that should get printed out during compilation and put the `#include` for the new file inside a `KOKKOS_ENABLE_DEPRECATED_CODE` guard.

william76 commented on src/graph/KokkosGraph_Distance2Color.hpp:
```diff
@@ -0,0 +1,309 @@
+
+#include "KokkosKernels_Utils.hpp"
+
+
```
@srajama1 what are the type conversions you're referencing?

srajama1 replied:
See an example in other kernels such as KokkosSparse_gauss_seidel.hpp. This is primarily for const vs non-const views.

william76 commented:
I've merged my updates in and am re-running some spot checks.

william76 commented:
I re-ran spot-check with the current updates:

## RHEL7
```
--------------------------------------------------------------------------------
-
-           K O K K O S   K E R N E L S   S P O T   C H E C K
-
--------------------------------------------------------------------------------

Kokkos
- URL...: git@github.com:william76/kokkos.git
- Branch: develop
- SHA1..: 86f10d250948f0a2f3c62eea1a4342f76b104ab8

KokkosKernels
- URL...: git@github.com:william76/kokkos-kernels.git
- Branch: Distance2GraphColor-WIP-v002
- SHA1..: fe04e608fcc8ebbcbb7fc0c808f5c96d5b6a4da0

Spot Check Command
/home/wcmclen/dev/kk-dev/source/KokkosKernels/scripts/test_all_sandia \
    --spot-check \
    --kokkoskernels-path=/home/wcmclen/dev/kk-dev/source/KokkosKernels \
    --kokkos-path=/home/wcmclen/dev/kk-dev/source/Kokkos \
    --arch="" \
    --skip-hwloc \
    gcc  \
    --num=3

ModuleCmd_Load.c(213):ERROR:105: Unable to locate a modulefile for 'sems-git'
Running on machine: sems
Going to test compilers:  gcc/5.3.0 gcc/7.2.0
Testing compiler gcc/5.3.0
Testing compiler gcc/7.2.0
  Starting job gcc-7.2.0-Serial-release
  Starting job gcc-5.3.0-Serial-release
  PASSED gcc-7.2.0-Serial-release
  PASSED gcc-5.3.0-Serial-release
#######################################################
PASSED TESTS
#######################################################
gcc-5.3.0-Serial-release build_time=169 run_time=2480
gcc-7.2.0-Serial-release build_time=158 run_time=2345
#######################################################
FAILED TESTS
#######################################################

--------------------------------------------------------------------------------
- Elapsed time: 2649 seconds.
0m0.011s 0m0.009s
111m33.797s 65m6.047s
--------------------------------------------------------------------------------
```

## White
```
--------------------------------------------------------------------------------
-
-           K O K K O S   K E R N E L S   S P O T   C H E C K
-
--------------------------------------------------------------------------------

Kokkos
- URL...: git@github.com:william76/kokkos.git
- Branch: develop
- SHA1..: 86f10d250948f0a2f3c62eea1a4342f76b104ab8

KokkosKernels
- URL...: git@github.com:william76/kokkos-kernels.git
- Branch: Distance2GraphColor-WIP-v002
- SHA1..: fe04e608fcc8ebbcbb7fc0c808f5c96d5b6a4da0

Spot Check Command
/home/wcmclen/dev/kk-dev/source/KokkosKernels/scripts/test_all_sandia \
    --spot-check \
    --kokkoskernels-path=/home/wcmclen/dev/kk-dev/source/KokkosKernels \
    --kokkos-path=/home/wcmclen/dev/kk-dev/source/Kokkos \
    --arch="Power8,Pascal60" \
    --skip-hwloc \
    cuda \
    gcc \
    --num=2

Running on machine: white
Going to test compilers:  cuda/9.2.88 cuda/10.0.130 gcc/6.4.0 gcc/7.2.0
Testing compiler cuda/9.2.88
Testing compiler cuda/10.0.130
  Starting job cuda-9.2.88-Cuda_OpenMP-release
  PASSED cuda-9.2.88-Cuda_OpenMP-release
Testing compiler gcc/6.4.0
  Starting job cuda-10.0.130-Cuda_Serial-release
  PASSED cuda-10.0.130-Cuda_Serial-release
Testing compiler gcc/7.2.0
  Starting job gcc-6.4.0-OpenMP_Serial-release
  PASSED gcc-6.4.0-OpenMP_Serial-release
  Starting job gcc-7.2.0-OpenMP-release
  PASSED gcc-7.2.0-OpenMP-release
  Starting job gcc-7.2.0-Serial-release
  PASSED gcc-7.2.0-Serial-release
  Starting job gcc-7.2.0-OpenMP_Serial-release
  PASSED gcc-7.2.0-OpenMP_Serial-release
#######################################################
PASSED TESTS
#######################################################
cuda-10.0.130-Cuda_Serial-release build_time=881 run_time=793
cuda-9.2.88-Cuda_OpenMP-release build_time=910 run_time=371
gcc-6.4.0-OpenMP_Serial-release build_time=352 run_time=904
gcc-7.2.0-OpenMP-release build_time=223 run_time=228
gcc-7.2.0-OpenMP_Serial-release build_time=336 run_time=895
gcc-7.2.0-Serial-release build_time=199 run_time=640
#######################################################
FAILED TESTS
#######################################################

--------------------------------------------------------------------------------
- Elapsed time: 6740 seconds.
0m0.000s 0m0.016s
1302m6.342s 25m57.814s
--------------------------------------------------------------------------------
```

srajama1 commented:
@william76 : Is this ready to be re-reviewed ?

william76 commented:
@srajama1 Yes. I think I had a couple questions left for you that I added into the comments you'd put in-line. I think the big thing is what to do with the experimental code to get the tight-bound on the distance-2 degree of a node. I did move the function that calls it out of the interface and into the Impl file so I can circle back around to them and see if we can get that d2-degree sped up. If we'd rather keep that out of this PR entirely, we can get this to the point we're ready to merge in with it and I'll split off a branch to save it and then remove it from what will be merged in... other than that, I think I addressed the other stuff we'd discussed.
